File: lua/intermediate/oop/inheritance.lua
-- inheritance.lua
-- Demonstrating inheritance hierarchies in Lua

-- Our class creation function with support for inheritance
local function create_class(base_class)
    local new_class = {}
    new_class.__index = new_class

    -- Set up inheritance
    if base_class then
        setmetatable(new_class, {
            __index = base_class     -- This allows method lookup in the base class
        })
    end

    -- Constructor that creates new instances
    function new_class:new(...)
        local instance = setmetatable({}, self)
        if instance.init then
            instance:init(...)
        end
        return instance
    end

    return new_class
end

-- Animal base class
local Animal = create_class()

function Animal:init(name)
    self.name = name or "unnamed"
    self.energy = 100
end

function Animal:eat(food)
    print(self.name .. " is eating " .. food)
    self.energy = self.energy + 25
    if self.energy > 100 then
        self.energy = 100
    end
end

function Animal:sleep(hours)
    print(self.name .. " is sleeping for " .. hours .. " hours")
    self.energy = self.energy + 10 * hours
    if self.energy > 100 then
        self.energy = 100
    end
end

function Animal:move()
    if self.energy >= 10 then
        print(self.name .. " is moving")
        self.energy = self.energy - 10
    else
        print(self.name .. " is too tired to move")
    end
end

function Animal:status()
    return self.name .. " has " .. self.energy .. "% energy"
end

-- Mammal inherits from Animal
local Mammal = create_class(Animal)

function Mammal:init(name, fur_color)
    Animal.init(self, name)     -- Call base class initializer
    self.fur_color = fur_color or "brown"
    self.body_temp = 37         -- Celsius
end

function Mammal:nurse()
    if self.energy >= 15 then
        print(self.name .. " is nursing their young")
        self.energy = self.energy - 15
    else
        print(self.name .. " is too tired to nurse")
    end
end

function Mammal:status()
    -- Override and extend the base method
    return Animal.status(self) .. " and " .. self.fur_color .. " fur"
end

-- Dog inherits from Mammal
local Dog = create_class(Mammal)

function Dog:init(name, fur_color, breed)
    Mammal.init(self, name, fur_color)     -- Call parent initializer
    self.breed = breed or "mixed"
    self.loyalty = 100
end

function Dog:bark()
    print(self.name .. " says: Woof!")
    self.energy = self.energy - 5
end

function Dog:fetch(item)
    if self.energy >= 20 then
        print(self.name .. " fetches the " .. item)
        self.energy = self.energy - 20
        self.loyalty = self.loyalty + 10
    else
        print(self.name .. " is too tired to fetch")
    end
end

function Dog:status()
    -- Override and extend parent methods
    return Mammal.status(self) .. ", " .. self.breed .. " breed, and " ..
        self.loyalty .. "% loyalty"
end

-- Bird inherits from Animal
local Bird = create_class(Animal)

function Bird:init(name, wing_span)
    Animal.init(self, name)
    self.wing_span = wing_span or 30     -- cm
    self.altitude = 0                    -- meters
end

function Bird:fly(meters)
    if self.energy >= meters / 10 then
        local energy_cost = meters / 10
        print(self.name .. " is flying " .. meters .. " meters")
        self.energy = self.energy - energy_cost
        self.altitude = 10     -- Simple simulation
    else
        print(self.name .. " is too tired to fly that far")
    end
end

function Bird:status()
    local altitude_desc = ""
    if self.altitude > 0 then
        altitude_desc = " flying at " .. self.altitude .. " meters"
    else
        altitude_desc = " on the ground"
    end
    return Animal.status(self) .. ", wingspan of " .. self.wing_span .. "cm" .. altitude_desc
end

-- Utility functions for testing inheritance
local function is_instance_of(instance, class)
    local mt = getmetatable(instance)

    while mt do
        if mt == class then
            return true
        end
        -- Move up the inheritance chain
        local parent_mt = getmetatable(mt)
        if parent_mt then
            mt = parent_mt.__index
        else
            return false
        end
    end

    return false
end

-- Create instances and test
print("\n--- Testing Base Animal Class ---")
local generic_animal = Animal:new("Generic Animal")
print(generic_animal:status())
generic_animal:eat("food")
generic_animal:move()
print(generic_animal:status())

print("\n--- Testing Mammal Inheritance ---")
local generic_mammal = Mammal:new("Generic Mammal", "black")
print(generic_mammal:status())
generic_mammal:eat("berries")     -- Inherited method
generic_mammal:nurse()            -- New method
print(generic_mammal:status())

print("\n--- Testing Dog Inheritance ---")
local rex = Dog:new("Rex", "golden", "Retriever")
print(rex:status())
rex:eat("dog food")     -- From Animal
rex:nurse()             -- From Mammal
rex:bark()              -- Dog-specific
rex:fetch("ball")       -- Dog-specific
print(rex:status())

print("\n--- Testing Bird Inheritance ---")
local tweety = Bird:new("Tweety", 20)
print(tweety:status())
tweety:eat("seeds")     -- From Animal
tweety:fly(100)         -- Bird-specific
print(tweety:status())

print("\n--- Testing Instance Relationships ---")
print("Is rex an Animal?", is_instance_of(rex, Animal))           -- true
print("Is rex a Mammal?", is_instance_of(rex, Mammal))            -- true
print("Is rex a Dog?", is_instance_of(rex, Dog))                  -- true
print("Is rex a Bird?", is_instance_of(rex, Bird))                -- false

print("Is tweety an Animal?", is_instance_of(tweety, Animal))     -- true
print("Is tweety a Bird?", is_instance_of(tweety, Bird))          -- true
print("Is tweety a Mammal?", is_instance_of(tweety, Mammal))      -- false

print("\n--- Testing Method Overriding ---")
print("Animal status: " .. generic_animal:status())
print("Mammal status: " .. generic_mammal:status())
print("Dog status: " .. rex:status())
print("Bird status: " .. tweety:status())

File: ruby/intermediate/oop/inheritance.rb
# Inheritance allows a class to inherit behavior from another class
# This creates an "is-a" relationship between classes

########################################################################
# Basic Inheritance
########################################################################

# Parent class (superclass, base class)
class Animal
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def speak
    "Some generic animal sound"
  end

  def sleep
    "#{@name} is sleeping."
  end

  def move
    "#{@name} is moving."
  end
end

# Child class (subclass, derived class) inherits from Animal
class Dog < Animal
  # Override the speak method
  def speak
    "Woof!"
  end

  # Add a new method specific to dogs
  def fetch
    "#{@name} is fetching the ball!"
  end
end

# Another child class
class Cat < Animal
  def speak
    "Meow!"
  end

  def purr
    "#{@name} is purring."
  end
end

# Create instances and test
dog = Dog.new("Rex")
cat = Cat.new("Whiskers")

puts "Dog:"
puts dog.name   # Rex (inherited from Animal)
puts dog.speak  # Woof! (overridden)
puts dog.sleep  # Rex is sleeping. (inherited)
puts dog.fetch  # Rex is fetching the ball! (Dog-specific)

puts "\nCat:"
puts cat.name   # Whiskers
puts cat.speak  # Meow! (overridden)
puts cat.purr   # Whiskers is purring. (Cat-specific)

# This would raise an error - cats can't fetch
# puts cat.fetch  # NoMethodError

########################################################################
# Inheritance Hierarchy and Method Lookup
########################################################################

# Ruby searches for methods up the inheritance chain
puts "\nInheritance hierarchy:"
puts Dog.ancestors.inspect  # [Dog, Animal, Object, Kernel, BasicObject]

class Vehicle
  def start_engine
    "Engine starting..."
  end
end

class Car < Vehicle
  def drive
    "Car is driving."
  end
end

class SportsCar < Car
  def drive
    "SportsCar is driving FAST!"
  end
end

sports_car = SportsCar.new
puts sports_car.drive         # SportsCar is driving FAST!
puts sports_car.start_engine  # Engine starting... (inherited from Vehicle)

########################################################################
# The super Keyword
########################################################################

class Parent
  def greet
    "Hello from Parent"
  end

  def introduce(name)
    "Hi #{name}, I'm a parent"
  end

  def report(*items)
    "Parent received: #{items.join(", ")}"
  end
end

class Child < Parent
  # Override greet and call parent's version using super
  def greet
    "#{super} and Child"
  end

  # Override introduce and pass specific arguments to super
  def introduce(name)
    "#{super(name)} and a child"
  end

  # Override report and pass all arguments to super
  def report(*items)
    "Child: #{super}"
  end
end

child = Child.new
puts child.greet      # Hello from Parent and Child
puts child.introduce("Alice")  # Hi Alice, I'm a parent and a child
puts child.report("apple", "banana")  # Child: Parent received: apple, banana

########################################################################
# Constructor Inheritance
########################################################################

class Person
  attr_reader :name, :age

  def initialize(name, age)
    @name = name
    @age = age
  end

  def introduce
    "Hi, I'm #{@name} and I'm #{@age} years old."
  end
end

class Student < Person
  attr_reader :university

  # Extend the parent constructor
  def initialize(name, age, university)
    # Call the parent constructor first
    super(name, age)
    # Add our own initialization
    @university = university
  end

  def introduce
    "#{super} I study at #{@university}."
  end
end

student = Student.new("Bob", 20, "Harvard")
puts student.introduce  # Hi, I'm Bob and I'm 20 years old. I study at Harvard.

########################################################################
# Multiple Inheritance (Ruby doesn't support this directly)
########################################################################

# Ruby doesn't support multiple inheritance, but uses modules for similar functionality
# (see modules.rb and mixins.rb for more)

########################################################################
# Inheritance vs. Composition
########################################################################

# Inheritance ("is-a" relationship) example:
# A SportsCar is a Car

# Composition ("has-a" relationship) example:
class Engine
  def start
    "Engine started"
  end

  def stop
    "Engine stopped"
  end
end

class ElectricCar
  # Composition: ElectricCar has an Engine
  def initialize
    @engine = Engine.new
  end

  def start_car
    "Electric car: #{@engine.start}"
  end

  def stop_car
    "Electric car: #{@engine.stop}"
  end
end

electric_car = ElectricCar.new
puts electric_car.start_car  # Electric car: Engine started
puts electric_car.stop_car   # Electric car: Engine stopped

########################################################################
# Method Visibility in Inheritance
########################################################################

class Base
  def public_method
    "This is public"
  end

  protected

  def protected_method
    "This is protected"
  end

  private

  def private_method
    "This is private"
  end
end

class Derived < Base
  def call_methods
    puts public_method       # Works - public methods are accessible anywhere
    puts protected_method    # Works - protected methods are accessible to subclasses
    puts private_method      # Works - private methods are accessible to subclasses in Ruby
  end

  def call_on_other(other_derived)
    puts other_derived.public_method        # Works - public
    puts other_derived.protected_method     # Works - protected methods can be called on other instances of the same class family
    # This would fail:
    # puts other_derived.private_method     # Error - private methods can't be called with an explicit receiver
  end
end

derived1 = Derived.new
derived2 = Derived.new
derived1.call_methods
derived1.call_on_other(derived2)

# This would fail:
# derived1.private_method    # Error - private method
